File: app/HDoc/h/src/Com/PathEntry.hs
-- Copyright 27-Jan-2020 ºDeme
-- GNU General Public License - V3 <http://www.gnu.org/licenses/>

--- Path entry data.

module Com.PathEntry
  ( T(..)
  , new
  , toJs
  , fromJs
  ) where

import qualified Dm.Js as Js
import Dm.Result

--- T
data T = T
          { name :: String
          , path :: String
          , selected :: Bool
          , exists :: Bool
          }

--- new name path
new :: String -> String -> T
new ename epath = T ename epath True False

--- toJs e
toJs :: T -> Js.T
toJs (T n p s e) = Js.wa $ [ Js.ws n, Js.ws p, Js.wb s, Js.wb e ]

--- fromJs js
fromJs :: Js.T -> Result T
fromJs js =
  Js.ra js >>=
  \a ->       Js.rs (a!!0) >>=
        \n -> Js.rs (a!!1) >>=
        \p -> Js.rb (a!!2) >>=
        \s -> Js.rb (a!!3) >>=
        \e -> Right $ T n p s e

File: app/YetiDoc/js/src/data/PathEntry.js
// Copyright 28-Feb-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

/**
    Application path entry.
**/

export default class PathEntry {

  /* .
  _rc_ PathEntry : from
  name : string
  path : string
  selected : boolean
  exists: boolean
  */

  /*--*/
  /**
      @param {string} name
      @param {string} path
      @param {boolean} selected
      @param {boolean} exists
  **/
  constructor (name, path, selected, exists) {

    /**
        @private
        @type {string}
    **/
    this._name = name;

    /**
        @private
        @type {string}
    **/
    this._path = path;

    /**
        @private
        @type {boolean}
    **/
    this._selected = selected;

    /**
        @private
        @type {boolean}
    **/
    this._exists = exists;

  }

  /**
      @return {string}
  **/
  get name () {
    return this._name;
  }

  /**
      @return {string}
  **/
  get path () {
    return this._path;
  }

  /**
      @return {boolean}
  **/
  get selected () {
    return this._selected;
  }

  /**
      @return {boolean}
  **/
  get exists () {
    return this._exists;
  }

  /**
      @param {!Array<?>} serial
      @return {!PathEntry}
  **/
  static fromJs (serial) {
    return new PathEntry(
      serial[0],
      serial[1],
      serial[2],
      serial[3]
    );
  }
  /*--*/
}
